frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx

```diff
@@ -0,0 +1,387 @@
+import {
+  getAlertManagerConfig,
+  patchAlertManagerConfig,
+  receiverTypes,
+} from '../alert-manager-utils';
+import { K8sResourceKind } from '../../../module/k8s';
+import {
+  AlertManagerConfig,
+  AlertManagerReceiver,
+  AlertManagerRoute,
+} from '../alert-manager-config';
+import { RoutingLabelEditor } from './routing-labels-editor';
+import * as PagerDutyForm from './pagerduty-receiver-form';
+import * as WebhookForm from './webhook-receiver-form';
+
+/**
+ * Converts routes of a specific Receiver:
+ * {
+ *   receiver: "MyReceiver",
+ *   match: {
+ *     severity: "warning",
+ *     cluster: "myCluster"
+ *   },
+ *   match_re: {
+ *    service: "$foobar"
+ *  }
+};
+ * ...to array of labels for Routing Labels Editor component
+ * [
+ *   {
+ *     "name": "severity",
+ *     "value": "warning",
+ *     "isRegex": false
+ *   },
+ *   {
+ *     "name": "cluster",
+ *     "value": "myCluster",
+ *     "isRegex": false
+ *   },
+ *   {
+ *     "name": "service",
+ *     "value": "$foobar",
+ *     "isRegex": true
+ *   }
+ * ]
+ */
+const getReceiverRouteLabels = (receiver: AlertManagerReceiver, routes: AlertManagerRoute[]) => {
+  const routesOfReceiver = _.find(
+    routes,
+    (aRoute: AlertManagerRoute) => aRoute.receiver === receiver.name,
+  );
+  const matches = _.map(_.get(routesOfReceiver, 'match', {}), (v, k) => {
+    return { name: k, value: v, isRegex: false };
+  });
+  const regexMatches = _.map(_.get(routesOfReceiver, 'match_re', {}), (v, k) => {
+    return { name: k, value: v, isRegex: true };
+  });
+  return _.concat([], matches, regexMatches);
+};
+
+/**
+ * Returns new Route object
+ * Ex:
+ * {
+ *   receiver: myNewReceiver,
+ *   match: {
+ *     "severity": "warning",
+ *     "cluster": "myCluster"
+ *   }
+ *   match_re {
+ *     "service": "^(foo1|foo2|baz)$",
+ *   }
+ * }
+ */
+const createRoute = (receiver: AlertManagerReceiver, routeLabels) => {
+  const newRoute = { receiver: receiver.name };
+  _.each(routeLabels, (label) => {
+    _.set(newRoute, [label.isRegex ? 'match_re' : 'match', label.name], label.value);
+  });
+  return newRoute;
+};
+
+/**
+ * Returns new Receiver object
+ * Ex:
+ * {
+ *   name: MyNewReceiver
+ *   pagerduty_configs: {
+ *     routing_key: <integration_key>
+ *   }
+ * }
+ */
+const createReceiver = (formValues, createReceiverConfig, receiverToEdit) => {
+  const receiverConfig = createReceiverConfig(
+    formValues,
+    receiverToEdit && receiverToEdit[formValues.receiverType]
+      ? receiverToEdit[formValues.receiverType][0] // pass in receiver config if editing existing receiver
+      : {},
+  );
+  return {
+    name: formValues.receiverName,
+    [formValues.receiverType]: [{ ...receiverConfig }],
+  };
+};
+
+const subFormFactory = (receiverType) => {
+  switch (receiverType) {
+    case 'pagerduty_configs':
+      return PagerDutyForm;
+    case 'webhook_configs':
+      return WebhookForm;
+    default:
+      return WebhookForm;
```

Comment:
Shouldn't we handle unrecognized types somehow?

Reply:
Not really.  an unknown receiver type would never get here, it is redirected to 'Edit YAML' instead of 'Edit Receiver' (via form).

Reply:
> Not really, an unknown receiver type would never get here, it is redirected to 'Edit YAML' instead of 'Edit Receiver' (via form).

If it's never expected to happen, even more reason to show an error!

Reply:
Ok, updated [here](https://github.com/openshift/console/pull/3937/commits/5a2398fe5b22feb28a7974e7da50bad2dea8ade0#diff-1f88ba8bd0ec27a0f637366e14fa9096R136)

Reply:
Actually, had to revert this.  When creating a new form, the receiverType is not know, which is ok, so we do not want to throw an error.